packages/dev/core/src/Meshes/Compression/dracoCompressionWorker.ts

```diff
@@ -1,48 +1,154 @@
+            if (options.quantizationBits && options.quantizationBits[attribute.dracoAttribute]) {
+                encoder.SetAttributeQuantization(encoderModule[attribute.dracoAttribute], options.quantizationBits[attribute.dracoAttribute]);
+            }
+        }
+
+        // Set the options
+        if (options.method) {
+            encoder.SetEncodingMethod(encoderModule[options.method]);
+        }
+        if (options.encodeSpeed !== undefined && options.decodeSpeed !== undefined) {
+            encoder.SetSpeedOptions(options.encodeSpeed, options.decodeSpeed);
+        }
+
+        // Encode to native buffer
+        encodedNativeBuffer = new encoderModule.DracoInt8Array() as DracoInt8Array;
+        const encodedLength = encoder.EncodeMeshToDracoBuffer(mesh, encodedNativeBuffer);
+        if (encodedLength <= 0) {
+            throw new Error("Draco encoding failed.");
+        }
+
+        // Copy the native buffer data to worker heap
+        // Reuse `destination` buffer if possible (which should always be the case for our default worker method).
+        const availableBytes = destination ? destination.byteLength : 0;
+        const buffer = availableBytes < encodedLength ? new ArrayBuffer(encodedLength) : destination!;
+        const encodedData = new Int8Array(buffer, 0, encodedLength);
+        for (let i = 0; i < encodedLength; i++) {
+            encodedData[i] = encodedNativeBuffer.GetValue(i);
+        }
+
+        return { data: encodedData, attributeIDs: attributeIDs };
+    } finally {
+        if (mesh) {
+            encoderModule.destroy(mesh);
+        }
+        if (meshBuilder) {
+            encoderModule.destroy(meshBuilder);
+        }
+        if (encoder) {
+            encoderModule.destroy(encoder);
+        }
+        if (encodedNativeBuffer) {
+            encoderModule.destroy(encodedNativeBuffer);
+        }
+    }
 }
 
-interface AttributeMessage extends AttributeData {
-    id: "attribute";
+/**
+ * The worker function that gets converted to a blob url to pass into a worker.
+ * To be used if a developer wants to create their own worker instance and inject it instead of using the default worker.
+ */
+export function EncoderWorkerFunction(): void {
+    let encoderPromise: PromiseLike<EncoderModule> | undefined;
+
+    onmessage = (event) => {
+        const message = event.data;
+        switch (message.id) {
+            case "init": {
+                // if URL is provided then load the script. Otherwise expect the script to be loaded already
+                if (message.url) {
+                    importScripts(message.url);
+                }
+                const initDecoderObject = message.wasmBinary ? { wasmBinary: message.wasmBinary } : {};
+                encoderPromise = DracoEncoderModule(initDecoderObject);
+                postMessage({ id: "initDone" });

```

Comment:
It's ok to pose this message before the `encoderPromise` completes?

Reply:
Yes, it should be alright. No encoding, which is triggered by "encodeMesh" message, will take place until the promise resolves. This is what the decoder worker function does.

But I also see no reason to _not_ wait for encoderPromise before posting "initDone"... might be nice to catch the error earlier than from "encodeMesh" handler.